pyrogram/client/client.py

```diff
@@ -571,26 +584,33 @@ def download_worker(self):
                     return None
 
                 if file_name is not None:
-                    path[0] = "downloads/{}".format(file_name)
+                    path[0] = os.path.join(download_directory, file_name)
 
                 try:
-                    os.remove("downloads/{}".format(file_name))
+                    os.remove(os.path.join(download_directory, file_name))
                 except OSError:
                     pass
                 finally:
                     try:
-                        os.renames("{}".format(tmp_file_name), "downloads/{}".format(file_name))
-                    except OSError:
-                        pass
+                        if download_directory:
+                            os.makedirs(download_directory, exist_ok=True)
+                        else:
+                            os.makedirs(os.path.dirname(file_name), exist_ok=True)
+
+                        # avoid errors moving between drives/partitions etc.
+                        shutil.move(tmp_file_name, os.path.join(download_directory, file_name))
+                    except OSError as e:
+                        log.error(e, exc_info=True)
             except Exception as e:
                 log.error(e, exc_info=True)
             finally:
                 done.set()
 
                 try:
-                    os.remove("{}".format(tmp_file_name))
-                except OSError:
-                    pass
+                    os.remove(tmp_file_name)
+                except OSError as e:
+                    if not e.errno == errno.ENOENT:
```

Comment:
Care to explain why is this necessary?

Reply:
The ``if not e.errno == errno.ENOENT`` is for logging an exception in the case that the file cannot be removed, but it is for some other reason aside from that it does not exist. Such as if it is still in use (at least on windows), which might be helpful in revealing a subtle bug if there is one.

It would be better off as ``if e.errno != errno.ENOENT`` actually, not sure why I wrote it with the not.

This could be worth doing elsewhere possibly.

Reply:
Using tempfile just made this whole thing overcomplicated in my opinion, now we need to handle file descriptors manually outside a context manager. a .temp file in the current working directory was doing no harm, we could also have make it hidden by prepending a dot eventually.

Reply:
The method was already generating unique file names, so I can't really see any benefit in using tempfile here

Reply:
> Using tempfile just made this whole thing overcomplicated in my opinion, now we need to handle file descriptors manually outside a context manager. a .temp file in the current working directory was doing no harm, we could also have made it hidden by prepending a dot eventually.

The FD should be closed inside **get_file** leaving just the unopened temp file in an OS specific location.  Users of **get_file** should never need to manually close an FD.

The "." is not what is causing the **file_name** parameter to be treated as a full path.  It is the presence of either "/" or "\" in the path, and no trailing directory separator, which causes it to treat it as a full file path.

using a **file_name** like this ".filename" should result in a hidden file being placed in the "downloads" folder of the current directory.  The "." is not interpreted as anything special.

Reply:
The idea of a OS provided tempfile better guarantees that the file can actually be created.

For instance, if I were to decide to run the client on windows out of the programs folder directory with the working directory where the script is, or otherwise run the client in a non writable working directory, than the temp file could not be written there.

Same goes for the default downloads folder, which is always getting created in the working directory.

Reply:
> Using tempfile just made this whole thing overcomplicated in my opinion, now we need to handle file descriptors manually outside a context manager. a .temp file in the current working directory was doing no harm, we could also have made it hidden by prepending a dot eventually.

I slightly misinterpreted what you were saying about having the FD outside of a with statement, above.  Yes I agree that is a bit precarious. 

Though the code could make use of:

```tempfile.NamedTemporaryFile('wb', delete=False)`` inside of a **with** context manager instead and be equivalent.